---
title: "IReNA_pipline"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{IReNA_pipline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# IReNA

<!-- badges: start -->
<!-- badges: end -->

IReNA (Integrated Regulatory Network Analysis) is to reconstruct regulatory networks through integrating scRNA-seq and ATAC-seq data. For more details of IReNA package, please see https://github.com/jiang-junyao/IReNA

## Citation
If you use IReNA package, please cite the following Science paper:Â https://science.sciencemag.org/content/370/6519/eabb8598.

## Installation

You can install the stable version of IReNA using:
```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("IReNA")
```

Or, you can install the development version of IReNA using:
```r
install.packages("devtools")
devtools::install_github("jiang-junyao/IReNA")
```

##Example

### Select DNA motif of transcription factors from TRANSFAC database
IReNA contains DNA motif datasets for four species (Homo sapiens, Mus musculus, Zebrafish and Chicken) derived from TRANSFAC 2018 database. Following codes are used to call  the motif dataset from TRANSFAC or  user-defined motif dataset which should have the same format as these from TRANSFAC database.
load IReNA package
```{r}
library(IReNA)
```
Call Mus musculus motif database
```{r}
motif1 = Tranfac201803_Mm_MotifTFsF
```
Call Homo sapiens motif database
```{r}
motif1 = Tranfac201803_Hs_MotifTFsF
```
Call Zebrafish motif database
```{r}
motif1 = Tranfac201803_Zf_MotifTFsF
```
Call Chicken motif database
```{r}
motif1 = Tranfac201803_Ch_MotifTFsF
```
First, we merge footprints whose distance is less than 4 and get fasta from each footprints based on reference genome through function get_merged_fasta(). Reference genome should be fasta/fa format, and you can download it from https://hgdownload.soe.ucsc.edu/downloads.html#alpaca or other genome database website.
```r
###merge footprints whose distance is less than 4
fdr005 <- read.table('mmATACPhxW_CuFiQ10No_sorted_fdr0.050000.bed',,sep='\t',header = T)
fastadir <- 'Genome/GRCm38Chr.fasta' 
merged_fasta <- get_merged_fasta(fdr005,fastadir)
write.table(merged_fasta,'merged_footprints.fasta',row.names=F,quote=F)
```
In this step, because Fimo software only have linux version, and it takes too long to implement the corresponding function on Windows, we generate a shell script to run Fimo software in shell. If you are familiar with linux system and Fimo, you can write your own commands as you like.
```r
Dir2 <- 'D:\\GIBH\\IReNA2 R package\\IReNA2\\ATAC\\outputdir'
find_motifs(motif1,step=20,Dir2,'merged_footprints.fasta')
### run the following commands in the shell
mv fimo_all.txt fimo_all.sh
chmod +x fimo_all.sh
sh ./fimo_all.sh
```
After get the result of Fimo, we can use [winscp](https://winscp.net/eng/download.php) or other related software to transfer fimo result files from linux to windos(If you use R in linux, please ignore this part). Then, we combine these Fimo consequence according to motif and motif Position weight matrix. Next, we load the peaks file and overlap differential peaks and motif footprints through overlap_footprints_peaks() function
```r
###Combine all footprints of motifs
combied <- combine_footprints(motif1,Dir2)
peaks <- read.delim('D:\\GIBH\\IReNA2 R package\\IReNA2\\ATAC\\Peaks\\mmATACPhxW_FcLog15Fdr05Diff.txt')
peak_bed <- get_bed(peaks)
overlapped <- overlap_footprints_peaks(combied,peak_bed)
```

Next, we intergrate bioconductor package [ChIPseeker](https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html) to get footprint-related genes. Before we run get_related_genes(), we need to specify TxDb, which can be download from: http://bioconductor.org/packages/release/BiocViews.html#___TxDb.
```r
###Merge and extend footprint regions
overlapped <- read.table('overlapped.txt')
###get footprint-related genes
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
txdb <- TxDb.Mmusculus.UCSC.mm10.knownGene
list1 <- get_related_genes(overlapped,txdb = txdb,motif=Tranfac201803_Mm_MotifTFsF,Species = 'Mm')
Get candidate genes/TFs-related peaks
list2 <- get_peaks_genes(list1,MmscRNA_PHx_Exp_NewF)
```
MmscRNA_PHx_Exp_NewF is expression profile, whose rownames should be gene ENSEMBEL ID, first column should be Symbol, second column should be group of gene, third column should be FoldChange.
```{r}
print(MmscRNA_PHx_Exp_NewF[1:5,1:5])
```
Count the cuts of each position in footrprints by wig_track().
```r
bamfilepath1 <- 'mmATACCtrW00R1_CuFiQ10No_sorted.bam'
bamfilepath2 <- 'mmATACCtrW00R2_CuFiQ10No_sorted.bam'
cuts1 <- wig_track(bamfilepath = bamfilepath1,bedfile = list2[[2]])
cuts2 <- wig_track(bamfilepath = bamfilepath2,bedfile = list2[[2]])
wig_list <- list(cuts1,cuts2)
```
Use these cuts to calculate the FOS of footprints to identify enriched TFs which determine the regulatory relationship.
```{r}
regulatory_relationships <- Footprints_FOS(wig_list,Candid,MmscRNA_PHx_Exp_NewF)
```
Use functions in GReNA to get regulatory networks for enriched TFs of each module and intramodular network. For more details of GReNA, please see https://github.com/jiang-junyao/GReNA
